### PR TITLE
Sanitise inputs

### PR DESCRIPTION
**Pull request by TomHanmer29**
Made it so users can't enter bad inputs (as far as I've tested), updated the table styling, made it so you can add multiple copies at once.

### CODE REVIEW

**Review by SleepyUser — APPROVED** [review]

File: Bookish/Views/Catalog/BookEntry.cshtml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 @model Bookish.Views.Catalog.BookEntry
 
 @{
-    Layout = null;
+    ViewBag.Title = "Book Entered";
```
Comment (emilypatsko) [inline Bookish/Views/Catalog/BookEntry.cshtml:5]: Any reason you're not using `ViewData['Title']`?

File: Bookish/Views/Catalog/BookList.cshtml [hunk 2]
```diff
@@ -7,17 +7,31 @@
...
-           <th>Title</th>
+<html lang="en">
+    <head>
+        <script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>
+        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-icons@1.8.3/font/bootstrap-icons.css">
+        <link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/1.12.1/css/jquery.dataTables.css">
+        <script type="text/javascript" charset="utf8" src="https://cdn.datatables.net/1.12.1/js/jquery.dataTables.js"></script>
+
+        <title>Book List</title>
+    </head>
```
Comment (emilypatsko) [inline Bookish/Views/Catalog/BookList.cshtml:18]: Weird indentation going on here

File: Bookish/Views/Catalog/BookList.cshtml [hunk 3]
```diff
@@ -7,17 +7,31 @@
...
+<body>
+<div>
+    <table class="display" style="width:100%" id="myTable">
```
Comment (emilypatsko) [inline Bookish/Views/Catalog/BookList.cshtml:21]: Wouldn't use inline styles - look into using the classes available in Bootstrap, docs are [here](https://getbootstrap.com/docs/5.1/getting-started/introduction/)

File: Bookish/Views/Catalog/BookList.cshtml [hunk 4]
```diff
@@ -27,13 +41,17 @@
                 <td>@b.ISBN</td>
                 <td>@b.DatePublished</td>
                 <td>@b.CopyList.Count</td>
+                <td><button class="btn-outline-info"><i class="bi bi-pencil"></i></button></td>
```
Comment (emilypatsko) [inline Bookish/Views/Catalog/BookList.cshtml:44]: Just like you've used classes here! (following on from my above comment about using Bootstrap)

File: Bookish/Views/Catalog/BookList.cshtml [hunk 5]
```diff
@@ -27,13 +41,17 @@
...
         }
+        </tbody>
     </table>
-</head>
-<body>
-<div>
-      
 </div>
 </body>
+        <script>
```
Comment (emilypatsko) [inline Bookish/Views/Catalog/BookList.cshtml:52]: What does this script do?

File: Bookish/Views/Catalog/BorrowerEntry.cshtml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+﻿@page
+@model Bookish.Views.Catalog.BorrowerEntry
+
+@{
+    ViewBag.Title = "Borrower Entered";
```
Comment (emilypatsko) [inline Bookish/Views/Catalog/BorrowerEntry.cshtml:5]: Same again, use `ViewData`

File: Bookish/Views/Catalog/BorrowerEntry.cshtml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+<!DOCTYPE html>
+
+<html>
+<head>
+    <title></title>
+</head>
+<body>
+<div>
+    <marquee>Book entered successfully</marquee>
```
Comment (emilypatsko) [inline Bookish/Views/Catalog/BorrowerEntry.cshtml:16]: Same again about `<marquee>`

File: Bookish/Views/Catalog/BorrowerEntry.cshtml.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+﻿using Microsoft.AspNetCore.Mvc.RazorPages;
+
+namespace Bookish.Views.Catalog;
+
+public class BorrowerEntry : PageModel
```
Comment (emilypatsko) [inline Bookish/Views/Catalog/BorrowerEntry.cshtml.cs:5]: I'm not sure what this file is for... but the actual filename seems incorrect

File: Bookish/Views/Home/Index.cshtml [hunk 9]
```diff
@@ -1,24 +1,37 @@
 ﻿@{
     ViewData["Title"] = "Home Page";
 }
-@using Bookish.Models
-@model BookInputModel
 
+@model DefaultInputModel
 <div class="text-center">
     <h1 class="display-4">Welcome</h1>
```
Comment (emilypatsko) [inline Bookish/Views/Home/Index.cshtml:7]: You can remove these placeholder bits if you like 😛

File: Bookish/Views/Home/Index.cshtml [hunk 10]
```diff
@@ -1,24 +1,37 @@
...
         <fieldset>
-            
             <legend>Enter a book:</legend>
-            @Html.TextBoxFor(m => m.BookName, new{@type="text", placeholder="Book name"})
-            @Html.TextBoxFor(m => m.ISBN, new{@type="text", placeholder="ISBN", pattern="[0-9]{13}", title = "Please enter a 13 digit number"})
-            @Html.TextBoxFor(m => m.Publisher, new{@type="text", placeholder="Publisher"})
-            @Html.TextBoxFor(m => m.DatePublished, new{@type="date", placeholder="Publish date"})
-            @Html.TextBoxFor(m => m.Author, new{@type="text", placeholder="Author"})
+            @Html.TextBoxFor(m => m.BookInput.BookName, new{@type="text", placeholder="Book name"})
+            @Html.TextBoxFor(m => m.BookInput.ISBN, new{@type="text", placeholder="ISBN", pattern="[0-9]{13}", title = "Please enter a 13 digit number"})
```
Comment (emilypatsko) [inline Bookish/Views/Home/Index.cshtml:15]: If you want to be strict, you could find the _real_ pattern for ISBN numbers

File: Bookish/Controllers/CatalogController.cs [hunk 11]
```diff
@@ -14,83 +14,108 @@ public CatalogController(ILogger<CatalogController> logger)
     }
     
     // GET
-    public IActionResult BookEntry(BookInputModel model)
+    public IActionResult BookEntry(DefaultInputModel model)
```
Comment (emilypatsko) [inline Bookish/Controllers/CatalogController.cs:17]: It looks like you're adding new data to the database in this method - in which case, it's not a `GET` endpoint

File: Bookish/Views/Home/Index.cshtml [hunk 12]
```diff
@@ -1,24 +1,37 @@
...
+            @Html.TextBoxFor(m => m.BookInput.Publisher, new{@type="text", placeholder="Publisher"})
+            @Html.TextBoxFor(m => m.BookInput.DatePublished, new{@type="date", placeholder="Publish date"})
+            @Html.TextBoxFor(m => m.BookInput.Author, new{@type="text", placeholder="Author", pattern="^[A-Za-z ,.'-]+$" })
+            @Html.TextBoxFor(m => m.BookInput.NewCopies, new{@type="number", placeholder="copies"})
             <input type="submit" value="Enter book info" />
 
         </fieldset>
     }
+    
+    @using (Html.BeginForm("BorrowerEntry", "Catalog", FormMethod.Get))
```
Comment (emilypatsko) [inline Bookish/Views/Home/Index.cshtml:25]: Same here - you're submitting data, so should probably be using a `POST` method

File: Bookish/Views/Shared/_Layout.cshtml [hunk 13]
```diff
@@ -44,7 +44,9 @@
             &copy; 2022 - Bookish - <a asp-area="" asp-controller="Home" asp-action="Privacy">Privacy</a>
         </div>
     </footer>
-    <script src="~/lib/jquery/dist/jquery.min.js"></script>
+    @*
```
Comment (emilypatsko) [inline Bookish/Views/Shared/_Layout.cshtml:47]: If you're not using this, I'd remove it - not nice to commit commented code

File: Bookish/Controllers/CatalogController.cs [hunk 14]
```diff
@@ -14,83 +14,108 @@ public CatalogController(ILogger<CatalogController> logger)
...
             {
-                ISBN = isbn,
-                Title = bookName,
-                Publisher = publisher,
-                AuthorID = author.AuthorID,
-                DatePublished = datePublished
-            };
-            context.Books.Add(book);
-            context.SaveChanges();
+                var author = new Author()
```
[the remark this answers is not in the document]
Comment (TomHanmer29) [reply Bookish/Controllers/CatalogController.cs:40]: This is fixed in a new branch that I haven't yet added a pull request for :)

File: Bookish/Views/Catalog/BookEntry.cshtml [hunk 15]
```diff
@@ -13,7 +13,7 @@
 </head>
 <body>
 <div>
-    <p>Book entered successfully</p>
+    <marquee>Book entered successfully</marquee>
```
[the remark this answers is not in the document]
Comment (TomHanmer29) [reply Bookish/Views/Catalog/BookEntry.cshtml:16]: These were kinda just here for fun, the plan was to remove them later but we thought we'd put a visual indicator of whether you entered an input for now :)

File: Bookish/Views/Catalog/BookEntry.cshtml (hunk 1 above)
Comment (TomHanmer29) [reply Bookish/Views/Catalog/BookEntry.cshtml:5]: It breaks it and I have no idea why! When I try and enter a new book with viewData["Title"] I get this error: 
image
but with the viewBag it seems to work.

File: Bookish/Views/Home/Index.cshtml (hunk 10 above)
Comment (TomHanmer29) [reply Bookish/Views/Home/Index.cshtml:15]: The real pattern has random hyphens in it, I will deal with this in a future branch

File: Bookish/Controllers/CatalogController.cs [hunk 16]
```diff
@@ -14,83 +14,108 @@ public CatalogController(ILogger<CatalogController> logger)
...
             {
-                AuthorSurname = "Sinha",
-                AuthorForename = "Sanjib",
-            };
-            
-            context.Authors.Add(author);
-            context.SaveChanges();
-            
-            var book = new Book()
+                //member already exists, do something
```
[the remark this answers is not in the document]
Comment (emilypatsko) [reply Bookish/Controllers/CatalogController.cs:73]: You can usually add a special type of comment for to-dos:
```
// TODO: Write the rest of your comment here
```

Just a helpful tip :)

File: Bookish/Views/Catalog/BookEntry.cshtml (hunk 1 above)
Comment (emilypatsko) [reply Bookish/Views/Catalog/BookEntry.cshtml]: Interesting, not seen this before! Something to look into...

File: Bookish/Controllers/CatalogController.cs [hunk 17]
```diff
@@ -13,84 +13,111 @@ public CatalogController(ILogger<CatalogController> logger)
         _logger = logger;
     }
     
-    // GET
-    public IActionResult BookEntry(BookInputModel model)
+    // [HttpPost]
```
Comment (emilypatsko) [inline Bookish/Controllers/CatalogController.cs:16]: Shouldn't be commented out, it's valid code and tells the controller that this method is for PUT requests

File: Bookish/Models/DefaultInputModel.cs [hunk 18]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Bookish.Models;
+
+public class BookOrBorrowerInputModel
```
Comment (emilypatsko) [inline Bookish/Models/DefaultInputModel.cs:3]: You've renamed the class, but not the file

File: Bookish/Models/BookViewModel.cs [hunk 19]
```diff
@@ -6,4 +6,5 @@ namespace Bookish.Models;
 public class BookViewModel
 {
     public List<Book> CatalogEntries;
+    public IQueryable<Book> EntryQuery;
```
Comment (emilypatsko) [inline Bookish/Models/BookViewModel.cs:9]: Is this used anywhere now? If not, the class doesn't need it

**Review by emilypatsko — CHANGES_REQUESTED** [review]
Nearly there - couple more small changes but then it'll probably be good to go,  I'll take one last look before you merge

File: Bookish/Views/Catalog/BookList.cshtml [hunk 20]
```diff
@@ -7,17 +7,31 @@
...
+<html lang="en">
+    <head>
+        <script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>
+        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-icons@1.8.3/font/bootstrap-icons.css">
+        <link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/1.12.1/css/jquery.dataTables.css">
+        <script type="text/javascript" charset="utf8" src="https://cdn.datatables.net/1.12.1/js/jquery.dataTables.js"></script>
+
+        <title>Book List</title>
+    </head>
+<body>
```
[the remark this answers is not in the document]
Comment (emilypatsko) [reply Bookish/Views/Catalog/BookList.cshtml:19]: So so nearly there! The `<head>` and `<body>` blocks should be indented once more, as they're nested within the `<html>` block. Like the `<script>` block at the bottom of the file :)

**Review by emilypatsko — CHANGES_REQUESTED** [review]
One tiny change!

**Review by emilypatsko — APPROVED** [review]

**Pull request merged**